### PR TITLE
Draft of C# server reflection

### PR DESCRIPTION
**Pull request by jtattermusch**
Implementation of C# server reflection service (will be available as a separate nuget just as Grpc.HealthCheck is). Also adding some unit tests for the functionality and tested manually that the server works well together with`grpc_cli` cmdline tool.

I am planning to further polish reflection.proto file because I am not fully satisfied with its layout, but it's probably better to do that in a followup PR (this one is already big enough).

Fixes https://github.com/grpc/grpc/issues/2566.

### CODE REVIEW

**Comment by jtattermusch** [comment]
Test failures seem unrelated (and C# tests have passed).

File: src/csharp/Grpc.Reflection/ReflectionServiceImpl.cs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+        ServerReflectionResponse CreateErrorResponse(StatusCode status, string message)
+        {
+            return new ServerReflectionResponse
+            {
+                ErrorResponse = new ErrorResponse { ErrorCode = (int) status, ErrorMessage = message }
+            };
+        }
+
+        void CollectTransitiveDependencies(FileDescriptor descriptor, HashSet<FileDescriptor> pool)
```
Comment (apolcyn) [inline src/csharp/Grpc.Reflection/ReflectionServiceImpl.cs]: Relies on proto FileDescriptor dependencies not allowed to be circular? Probably is well known, but maybe comment/assert for it?

Comment (jtattermusch) [inline src/csharp/Grpc.Reflection/ReflectionServiceImpl.cs]: added a comment.

File: src/csharp/Grpc.Reflection/project.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    "owners": [ "grpc-packages" ],
+    "licenseUrl": "https://github.com/grpc/grpc/blob/master/LICENSE",
+    "projectUrl": "https://github.com/grpc/grpc",
+    "requireLicenseAcceptance": false,
+    "tags": [ "gRPC reflection" ]
+  },
+  "buildOptions": {
+    "define": [ "SIGNED" ],
+    "keyFile": "../keys/Grpc.snk",
+    "publicSign": true,
```
Comment (jtattermusch) [inline src/csharp/Grpc.Reflection/project.json]: I regenerated protect files and it's gone now.

**Comment by jtattermusch** [comment]
Addressed all the comments.

File: src/csharp/Grpc.Reflection.Tests/SymbolRegistryTest.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        public void FileByName()
+        {
+            Assert.AreSame(Google.Protobuf.WellKnownTypes.Duration.Descriptor.File, registry.FileByName("google/protobuf/duration.proto"));
+            Assert.IsNull(registry.FileByName("somepackage/nonexistent.proto"));
+        }
+
+        [Test]
+        public void FileContainingSymbol()
+        {
+            Assert.AreSame(Google.Protobuf.WellKnownTypes.Duration.Descriptor.File, registry.FileContainingSymbol("google.protobuf.Duration"));
```
Comment (apolcyn) [inline src/csharp/Grpc.Reflection.Tests/SymbolRegistryTest.cs]: `FileContainingSymbol` returns `Google.Protobuf.WellKnownTypes.Duration.Descriptor.File` for one proto, but `ReflectionReflection.Descriptor` for another, I'm a bit unclear here but these are both just file descriptor protos but under different field names?

Comment (jtattermusch) [inline src/csharp/Grpc.Reflection.Tests/SymbolRegistryTest.cs]: Yes. 
Explanation: FooReflection is a class generated for every .proto file (and provides access to the file desriptor).
Alternative way is to access the message descriptor for a message type (Duration.Descriptor is a message descriptor) and lookup the file descriptor that contains is (the `.File` part)

File: templates/src/csharp/Grpc.Reflection/project.json.template [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      "owners": [ "grpc-packages" ],
+      "licenseUrl": "https://github.com/grpc/grpc/blob/master/LICENSE",
+      "projectUrl": "https://github.com/grpc/grpc",
+      "requireLicenseAcceptance": false,
+      "tags": [ "gRPC reflection" ]
+    },
+    "buildOptions": {
+      "define": [ "SIGNED" ],
+      "keyFile": "../keys/Grpc.snk",
+      "publicSign": true,
```
Comment (jtattermusch) [inline templates/src/csharp/Grpc.Reflection/project.json.template]: Ah, you're right. Thanks.

**Comment by jadekler** [comment]
Hi @jtattermusch. Is there a tutorial describing how to implement C# server reflection? e.g. https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#enable-server-reflection or https://github.com/grpc/grpc-java/blob/master/documentation/server-reflection-tutorial.md#enable-server-reflection.

If not, could we make a brief one showing how to do this? Happy to file an issue if you prefer.

**Comment by jtattermusch** [comment]
@jadekler  we currently don't have such tutorial (I agree having one would be useful), but there's s unit test that demonstrates how to setup things (The key difference between java and C# is that in C# you need to manually register the services' Descriptor with the ReflectionServiceImpl):
https://github.com/grpc/grpc/blob/cdfc58b1f0a9475287ec52b38b41cf3cc5af6ed8/src/csharp/Grpc.Reflection.Tests/ReflectionClientServerTest.cs#L44

**Comment by jadekler** [comment]
Thanks @jtattermusch. A bit more context - grpc server reflection and [grpcurl](https://github.com/fullstorydev/grpcurl) will be presented at a large conference (Gophercon) next month. In preparation for that I'd love for all our docs to be in order. Would it be possible to create a short C# tutorial for me to link to [here](https://github.com/grpc/grpc/pull/16018)?

**Pull request merged**